W-13789629 create pdk project gr (#350)
mulesoft/docs-gateway: gateway/1.5/modules/ROOT/pages/policies-pdk-create-project.adoc

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

To begin developing a new custom policy with the Policy Development Kit (PDK), generate a new custom policy project.

== Before You Begin

Before you can create a custom policy project, ensure that you have installed all xref:policies-pdk-prequisites.adoc[PDK prequisites].

== Create a New Policy Project

To create a custom policy project, run the PDK CLI `create` command, replacing `<my-custom-policy>` with the name of your new policy:

[source,ssh]
----
anypoint-cli-v4 pdk policy-project create --name <my-custom-policy>
----

Running the `create` command creates a custom policy project with the specified name. The project appears in a new directory or a directory specified using the `--output-dir` flag. The following tutorials refer to this as your project's root directory.

> The `assetID` of the new policy is the policy's name in lower case with the spaces amended together with a `-`. For example, a policy with the name `My Header Injection Policy` has the `asstID` of `my-header-injection-policy`. 

Reviewer: there's still a typo here: "asstID" => "assetID"

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

To begin developing a new custom policy with the Policy Development Kit (PDK), generate a new custom policy project.

== Before You Begin

Before you can create a custom policy project, ensure that you have installed all xref:policies-pdk-prequisites.adoc[PDK prequisites].

== Create a New Policy Project

To create a custom policy project, run the PDK CLI `create` command, replacing `<my-custom-policy>` with the name of your new policy:

[source,ssh]
----
anypoint-cli-v4 pdk policy-project create --name <my-custom-policy>
----

> Running the `create` command creates a custom policy project with the specified name. The project appears in a new directory or a directory specified using the `--output-di` flag. The following tutorials refer to this as your project's root directory.

Reviewer: ```suggestion
Running the `create` command creates a custom policy project with the specified name. The project appears in a new directory or a directory specified using the `--output-dir` flag. The following tutorials refer to this as your project's root directory.
```

> The `assetID` of the new policy is the policy's name in lower case with the spaces amended together with a `-`. For example, a policy with the name `My Header Injection Policy` asstID is `my-header-injection-policy`. 

Reviewer: ```suggestion
The `assetID` of the new policy is the policy's name in lower case with the spaces amended together with a `-`. For example, a policy with the name `My Header Injection Policy`, `assetID` is `my-header-injection-policy`. 
```

> The `groupID` of the new policy is `groupID` of the organization configured in Anypoint CLI.

Reviewer: I would say:
The `groupID` of the new policy is `ID` of the organization configured in Anypoint CLI.

GroupID is an Exchange concept, which in most cases uses the ID of the organization that owns the asset.

├─ Cargo.toml   # Rust language standard file that contains the metadata required to build the policy
└─.project.yaml # Project manifest used by PDK
----

=== target/ and definition/target/

There are two separate `target` directories:

* `definition/target/`: contains the policy asset definition published to exchange
* `target/`: Contains the files for the Policy asset implementation and the output of the Rust compilation process

Both `target` directories and their content are included in the `.gitignore` file. You must compile the targets on each individual devices. For example, if you are developing a policy as part of a team, your team can publish the policy and it's source code to a remote repository, but each team member must build the target on their device for local testing.

=== definition/

The `definition` directory contains the `gcl.yaml` file that defines the policy's metadata and the configuration properties.

To configure the `gcl.yaml` file, see xref:policies-pdk-create-schema-definition.adoc[].

=== Cargo.toml

Custom policies for Flex Gateway are developed in the Rust Programming language and compiled to a WebAssembly binary. 

> The `Cargo.toml` is Rust project manifest. Its `[package]` section contains the basic information of the policy implementation. The version specified in `Cargo.toml` matches the policy asset version. You can manage the policy version by editing the version number in `Cargo.toml`. The version follows the Semantic Versionsing (SemVer) format: `<major-version>.<minor-version>.<patch-version>`.

Reviewer: > The version follows the Semantic Versionsing (SemVer) format: `<major-version>.<minor-version>.<patch-version>`

This is not entirely accurate. SemVer permits some formats we do not support, for example: `1.0.1-integration`. The previous version is valid in SemVer, but not valid for us. We only admit the 3 numbered parts of the version: `1.0.1`, `1.2.3`, etc... So we accept a subset of SemVer format.

=== src/

The `src` directory contains the policy's source code:

* `lib.rs`: Contains a set of methods that implement the behavior of the policy. These methods are the API our development kit offers to implement custom policies behavior. 

* `generated/config.rs` and `generated/mod.rs`: Are auto generated modules that make the policy configuration, as defined in the `gcl.yaml`, available for use in the lib.rs and any of the policy source files.

=== Makefile

This file contains all the callable actions, `make` commands, required for the policy development lifecycle:

* `make setup`: Installs the PDK internal dependencies for the rest of the Makefile goals. 
+
You ran this command in <<setup-a-new-policy-project>>.
* `make build-asset-files`: Generates all the policy asset files required to build, execute, and publish the policy. This command also updates the `config.rs` source code file with the latest configurations defined in the `gcl.yaml` file. 
+

> For more information about compiling a policy, see xref:policies-pdk-create-schema-definition.ado[].

Reviewer: This text is out of sync with the cited link

To begin developing a new custom policy with the Policy Development Kit (PDK), generate a new custom policy project.

To create a new project, complete the following tasks:

. <<before-you-begin>>
. <<create-a-new-policy-project>>
. <<setup-a-new-policy-project>>
. Review <<project-structure>>

== Before You Begin

Before you can create a custom policy project, ensure that you have installed all xref:policies-pdk-prequisites.adoc[PDK prequisites].

== Create a New Policy Project

To create a custom policy project, run the PDK CLI `create` command, replacing `<my-custom-policy>` with the name of your new policy:

[source,ssh]
----
anypoint-cli-v4 pdk policy-project create --name <my-custom-policy>
----

Running the `create` command creates a custom policy project with the specified name. The project appears in a new directory or a directory specified using the `--output-dir` flag. The following tutorials refer to this as your project's root directory.

The `assetID` of the new policy is the policy's name in lower case with the spaces amended together with a `-`. For example, a policy with the name `My Header Injection Policy` has the `assetID` of `my-header-injection-policy`. 

> The `groupID` of the new policy is ID of the organization configured in Anypoint CLI.

Reviewer 2: ```suggestion
The `groupID` of the new policy is the ID of the organization configured in Anypoint CLI.
```

> Additionally, the `create` command supports the following flags:

Reviewer 2: ```suggestion
The `create` command supports the following flags:
```

[%header%autowidth.spread,cols="a,a,a,a"]
|===
| Flag | Default Value | Required | Description

> | `--name=<value>` or `-n` | N/A | Yes | Name of the new policy
> | `--description=<value>` or `-d` | N/A | No | Description of the new policy

Reviewer 2: ```suggestion
| `--name=<value>` or `-n` | N/A | Yes | Name of the new policy.
| `--description=<value>` or `-d` | N/A | No | Description of the new policy.
```
For consistency with the other items in the table.

| `--output-dir=<value>` of `-o` | `assetID` | No | Directory the new policy is created in. +
If the directory does not exist, it is created. If no directory is specified, the directory name is the policy `assetID`.
| `--version=<value>` | `1.0.0` | No | Version of the new policy. +
The version format is `<major-release>.<minor-release>.<patch-release>`.
|===

== Setup the PDK Build Environment

> If you haven't previously setup your PDK build environment, you must run the `setup` command to install the PDK project's dependencies.

Reviewer 2: ```suggestion
If you haven't set up your PDK build environment, run the `setup` command to install the PDK project's dependencies.
```

> To setup your PDK environment:

Reviewer 2: ```suggestion
To set up your PDK environment:
```

> . Execute the following `curl` command to retrive your connected app bearer token. Replace your connected app values:

Reviewer 2: ```suggestion
. Execute the following `curl` command to retrieve your Connected App bearer token. Replace your Connected App values:
```

+
[source,ssh]
----
curl --location --request POST https://anypoint.mulesoft.com/accounts/api/v2/oauth2/token \
--data-urlencode 'client_id=<client_id>' \
--data-urlencode 'client_secret=<client_secret>' \
--data-urlencode 'grant_type=client_credentials'
----

. Save the returned `access_token`.

. Run the `setup` command:
+
[source,ssh]
----
make setup
----

> . Enter an `username` value.

Reviewer 2: ```suggestion
. Enter a `username` value.
```

. Enter the saved `access_token` when prompted for a `password`.

== Project Structure

After running the setup command, the policy's root directory contains the following directories and files shown in the directory tree: 

----
├─ definition/  # Contains the files that describe the Policy Definition
│  ├── target/  # Contains the build definition asset files
│  └── gcl.yaml # GCL file managed by the developer that describes the Policy Definition
├─ target/      # Contains the implementation asset files and the output of the policy build
├─ src/         # Contains the implementation source code
├─ test/        # Contains the artifacts required for running the policy locally
├─ Makefile     # Provides all the callable actions required during the policy development lifecycle
├─ Cargo.toml   # Rust language standard file that contains the metadata required to build the policy
└─.project.yaml # Project manifest used by PDK
----

=== target/ and definition/target/

There are two separate `target` directories:

> * `definition/target/`: contains the policy asset definition published to exchange
> * `target/`: Contains the files for the Policy asset implementation and the output of the Rust compilation process

Reviewer 2: ```suggestion
* `definition/target/`: Contains the policy asset definition published to Exchange.
* `target/`: Contains the files for the Policy asset implementation and the output of the Rust compilation process.
```

Both `target` directories and their content are included in the `.gitignore` file. You must compile the targets on each individual devices. For example, if you are developing a policy as part of a team, your team can publish the policy and it's source code to a remote repository, but each team member must build the target on their device for local testing.

=== definition/

The `definition` directory contains the `gcl.yaml` file that defines the policy's metadata and the configuration properties.

To configure the `gcl.yaml` file, see xref:policies-pdk-create-schema-definition.adoc[].

=== Cargo.toml

Custom policies for Flex Gateway are developed in the Rust Programming language and compiled to a WebAssembly binary. 

> The `Cargo.toml` is Rust project manifest. Its `[package]` section contains the basic information of the policy implementation. The version specified in `Cargo.toml` matches the policy asset version. You can manage the policy version by editing the version number in `Cargo.toml`. The version follows the format `<major-version>.<minor-version>.<patch-version>`.

Reviewer 2: ```suggestion
The `Cargo.toml` is the Rust project manifest. Its `[package]` section contains the basic information on the policy implementation. The version specified in `Cargo.toml` matches the policy asset version. You can manage the policy version by editing the version number in `Cargo.toml`. The version follows the format `<major-version>.<minor-version>.<patch-version>`.
```

Author: Changed 'on' to 'about'

=== src/

The `src` directory contains the policy's source code:

* `lib.rs`: Contains a set of methods that implement the behavior of the policy. These methods are the API our development kit offers to implement custom policies behavior. 

> * `generated/config.rs` and `generated/mod.rs`: Are auto generated modules that make the policy configuration, as defined in the `gcl.yaml`, available for use in the lib.rs and any of the policy source files.

Reviewer 2: ```suggestion
* `generated/config.rs` and `generated/mod.rs`: Auto-generated modules that make the policy configuration, as defined in the `gcl.yaml`, available for use in the `lib.rs` and any of the policy source files.
```

=== Makefile

> This file contains all the callable actions, `make` commands, required for the policy development lifecycle. Execute all `make` commands at the root level of you root directory:

Reviewer 2: ```suggestion
This file contains all the callable actions, `make` commands, required for the policy development lifecycle. Execute all `make` commands at the root level of your root directory:
```

* `make setup`: Installs the PDK internal dependencies for the rest of the Makefile goals. 
+

> You ran this command in <<setup-a-new-policy-project>>.

Reviewer 2: ```suggestion
You run this command in <<setup-a-new-policy-project>>.
```
Please check if this is correct

* `make build-asset-files`: Generates all the policy asset files required to build, execute, and publish the policy. This command also updates the `config.rs` source code file with the latest configurations defined in the `gcl.yaml` file. 
+
For more information about compiling a policy, see xref:policies-pdk-compile-policies.adoc[].
+
* `make build`: Runs the `make build-asset-files` command and compiles the WebAssembly binary of the policy. 
+
For more information about compiling a policy, see xref:policies-pdk-compile-policies.adoc[]. 

> * `make run`: Provides a simple way to execute the current build of the policy in Docker containerized environment. 

Reviewer 2: ```suggestion
* `make run`: Provides a simple way to execute the current build of the policy in a Docker containerized environment. 
```
